core/src/main/java/jenkins/scm/RunWithSCMMixIn.java

```diff
@@ -0,0 +1,161 @@
+                if (pr != null && pr.isWorseThan(Result.SUCCESS)) {
+                    // we are still building, so this is just the current latest information,
+                    // but we seems to be failing so far, so inherit culprits from the previous build.
+                    // isBuilding() check is to avoid recursion when loading data from old Hudson, which doesn't record
+                    // this information
+                    r.addAll(p.getCulprits());
+                }
+            }
+            for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+                for (ChangeLogSet.Entry e : c)
+                    r.add(e.getAuthor());
+            }
+
+            if (p instanceof AbstractBuild && upstreamCulprits) {
+                // If we have dependencies since the last successful build, add their authors to our list
+                if (p.getPreviousNotFailedBuild() != null) {
+                    Map<AbstractProject, AbstractBuild.DependencyChange> depmap =
+                            ((AbstractBuild<?,?>) p).getDependencyChanges((AbstractBuild<?,?>)p.getPreviousSuccessfulBuild());
+                    for (AbstractBuild.DependencyChange dep : depmap.values()) {
+                        for (AbstractBuild<?, ?> b : dep.getBuilds()) {
+                            for (ChangeLogSet.Entry entry : b.getChangeSet()) {
+                                r.add(entry.getAuthor());
+                            }
+                        }
+                    }
+                }
+            }
+
+            return r;
+        }
+
+        return new AbstractSet<User>() {
+            public Iterator<User> iterator() {
+                return new AdaptedIterator<String,User>(asRun().getCulpritIds().iterator()) {
+                    protected User adapt(String id) {
+                        return User.get(id);
+                    }
+                };
+            }
+
+            public int size() {
+                return asRun().getCulpritIds().size();
+            }
+        };
+    }
+
+    /**
+     * Returns true if this user has made a commit to this build.
+     */
+    public boolean hasParticipant(User user) {
+        for (ChangeLogSet<? extends ChangeLogSet.Entry> c : getChangeSets()) {
+            for (ChangeLogSet.Entry e : c)
+                try {
+                    if (e.getAuthor() == user)
+                        return true;
+                } catch (RuntimeException re) {
+                    LOGGER.log(Level.INFO, "Failed to determine author of changelog " + e.getCommitId() + "for " + asRun().getParent().getDisplayName() + ", " + asRun().getDisplayName(), re);
+                }
+        }
+        return false;
+    }
+
+    public interface RunWithSCM<JobT extends Job<JobT, RunT> & Queue.Task,
+            RunT extends Run<JobT, RunT> & RunWithSCM<JobT,RunT> & Queue.Executable> {
+        @Nonnull
+        List<ChangeLogSet<? extends ChangeLogSet.Entry>> getChangeSets();
```

Comment:
I was hoping we could take advantage of `-source 8` to have all the methods here other than `getRunWithSCMMixIn` get default implementations, and started prototyping doing this with other mixin classes. Unfortunately it does not seem to work in general (for example for `LazyLoadingJob._getRuns()`), because apparently a concrete class cannot pick up a default method when the method is defined in an abstract superclass. Thus

```java
public class Demo {
    public static void main(String[] args) {
        System.out.println(new C().answer());
    }
    public static abstract class AC {
        public abstract int x();
    }
    public interface I {
        default int answer() {return 42;}
    }
    public static class C extends AC implements I {}
}
```

fails to compile:

```
Demo.java:11: error: C is not abstract and does not override abstract method x() in AC
    public static class C extends AC implements I {}
1 error
```

However I think it _could_ work here, since the methods you are expecting these `Run`s to define are not declared in `Run`. I am still looking for possible uses in `ParameterizedJobMixIn` & `LazyBuildMixIn`.